pkg/logical-cluster-manager/cluster-manager-controller.go

```diff
@@ -0,0 +1,244 @@
+package clustermanager
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"time"
+
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/logical-cluster"
+
+	clusterproviderclient "github.com/kcp-dev/edge-mc/cluster-provider-client"
+	cluster "github.com/kcp-dev/edge-mc/cluster-provider-client/cluster"
+	v1alpha1apis "github.com/kcp-dev/edge-mc/pkg/apis/edge/v1alpha1"
+	edgeclient "github.com/kcp-dev/edge-mc/pkg/client/clientset/versioned"
+	"github.com/kcp-dev/edge-mc/pkg/client/informers/externalversions/edge/v1alpha1"
+)
+
+const controllerName = "cluster-manager"
+
+type op string
+
+const (
+	opAdd    op = "Add"
+	opUpdate op = "Update"
+	opDelete op = "Delete"
+)
+
+type queueItem struct {
+	op  op
+	key any
+}
```

Comment:
This is a very unusual controller design.  The usual controller design puts into the work queue only object references.  And I do not mean pointers, I mean namespace (if applicable) and name.  That was the workqueue guarantees that only one worker is working on a given object at a time.  If you put more fine grained information in the work queue entries then (a) it may not be the latest by the time it is dequeued and processed and (b) it can cause concurrent work to be done on the same object.